### PR TITLE
Starfield

### PR DESCRIPTION
**Pull request by nikitalita**
WIP, although everything is implemented now, nothing has been really tested yet.

Do not have a configuration flag for turning on Starfield stuff yet, it's just on by default right now, will add that later.

### CODE REVIEW

File: Caprica/papyrus/parser/PapyrusParser.cpp [hunk 1]
```diff
@@ -457,6 +467,46 @@ statements::PapyrusStatement* PapyrusParser::parseStatement(PapyrusFunction* fun
       return ret;
     }
 
+    case TokenType::kLock:
+    {
+      auto ret = alloc->make<statements::PapyrusLockStatement>(consumeLocation());
+
+      if (cur.type == TokenType::EOL) {
```
Comment (nikitalita) [inline Caprica/papyrus/parser/PapyrusParser.cpp:474]: should just be checking if the next token type is an identifier

File: Caprica/papyrus/parser/PapyrusParser.cpp [hunk 2]
```diff
@@ -457,6 +467,46 @@ statements::PapyrusStatement* PapyrusParser::parseStatement(PapyrusFunction* fun
...
+      expectConsumeEOLs();
+      while (!maybeConsume(TokenType::kEndLock))
+        ret->body.push_back(parseStatement(func));
+      expectConsumeEOLs();
+      return ret;
+    }
+    case TokenType::kTryLock:
+    {
+      auto ret = alloc->make<statements::PapyrusTryLockStatement>(consumeLocation());
+      if (cur.type == TokenType::EOL) {
```
Comment (nikitalita) [inline Caprica/papyrus/parser/PapyrusParser.cpp:493]: same here

File: Caprica/papyrus/PapyrusResolutionContext.cpp [hunk 3]
```diff
@@ -478,6 +493,8 @@ PapyrusIdentifier PapyrusResolutionContext::tryResolveFunctionIdentifier(const P
       fk = PapyrusBuiltinArrayFunctionKind::Remove;
     } else if (idEq(ident.res.name, "removelast")) {
       fk = PapyrusBuiltinArrayFunctionKind::RemoveLast;
+    } else if (idEq(ident.res.name, "getmatchingstructs")) {
```
Comment (nikitalita) [inline Caprica/papyrus/PapyrusResolutionContext.cpp:496]: add comment about this being unverified syntax

File: Caprica/papyrus/expressions/PapyrusFunctionCallExpression.cpp [hunk 4]
```diff
@@ -294,6 +305,38 @@ void PapyrusFunctionCallExpression::semantic(PapyrusResolutionContext* ctx, Papy
       case PapyrusBuiltinArrayFunctionKind::RemoveLast:
         getArgs("RemoveLast", 0);
         return;
+      case PapyrusBuiltinArrayFunctionKind::GetMatchingStructs:
+      {
+        getArgs("GetMatchingStructs", 3, 4);
```
Comment (nikitalita) [inline Caprica/papyrus/expressions/PapyrusFunctionCallExpression.cpp:310]: add comment about this being unverified syntax

File: Caprica/common/CapricaReportingContext.h [hunk 5]
```diff
@@ -72,6 +86,31 @@ NEVER_INLINE void warning_W##num##_##id(CapricaFileLocation location, arg1Type a
...
+  if (!m_Warn_##num##_##id_emitted) { \
+    warning(location, num, "%s\n\tFurther %d warnings will be suppressed.", msg, num); \
+    m_Warn_##num##_##id_emitted = true; \
+  } \
+}
+#define DEFINE_WARNING_ONCE_A1(num, id, msg, arg1Type, arg1Name) \
+static inline bool m_Warn_##num##_##id_emitted{false}; \
+NEVER_INLINE void warning_W##num##_##id(CapricaFileLocation location, arg1Type arg1Name) { \
+  if (!m_Warn_##num##_##id_emitted) { \
+    warning(location, num, "%s\n\tFurther %d warnings will be suppressed.", formatString(msg, arg1Name), num); \
```
Comment (nikitalita) [inline Caprica/common/CapricaReportingContext.h:101]:
```suggestion
    warning(location, num, "%s\n\tFurther %d warnings will be suppressed.", formatString(msg, arg1Name).c_str(), num); \
```

**Comment by nikitalita** [comment]
Ok, this should be good to go now besides the outstanding comment above. I have compiler game configuration implemented on another branch, which includes Skyrim support, but I’ll put that in another PR.

File: Caprica/papyrus/parser/PapyrusParser.cpp [hunk 6]
```diff
@@ -1127,6 +1184,7 @@ PapyrusUserFlags PapyrusParser::maybeConsumeUserFlags(CapricaUserFlagsDefinition
         break;
       case TokenType::kAutoReadOnly:
         flags.isAutoReadOnly = true;
+        flags.isAuto = true; // TODO: Orvid, do you want this to be true here too, or do you want the `IsAuto() && IsAutoReadOnly()` check in PapyrusProperty to be only `IsAutoReadOnly()`?
```
[the remark this answers is not in the document]
Comment (Orvid) [reply Caprica/papyrus/parser/PapyrusParser.cpp:1187]: This is the correct fix.

[the remark this answers is not in the document]
Comment (nikitalita) [reply Caprica/papyrus/parser/PapyrusParser.cpp:1187]:
```suggestion
        flags.isAuto = true;
```

**Pull request merged**